htlcswitch/circuit.go

```diff
@@ -53,82 +80,220 @@ func newPaymentCircuit(src, dest lnwire.ShortChannelID, key circuitKey,
+	if _, err := w.Write(destBytes[:]); err != nil {
+		return err
+	}
+
+	var code [1]byte
+	switch c.Obfuscator.(type) {
+	case *FailureObfuscator:
+		code[0] = failureObfuscatorCode
+	case *mockObfuscator:
+		code[0] = mockObfuscatorCode
+	default:
+		return errors.New("unknown obfuscator type")
+	}
+
+	if _, err := w.Write(code[:]); err != nil {
+		return err
+	}
+
+	return c.Obfuscator.Encode(w)
 }
 
-// circuitMap is a data structure that implements thread safe storage of
+// Key uniquely identifies an active circuit between two open channels.
+// Currently, the payment hash is used to uniquely identify each circuit.
+// TODO(roasbeef): include dest+src+amt in key
+func (c *paymentCircuit) Key() circuitKey {
+	return circuitKey(c.PaymentHash)
+}
+
+// circuitStorage is a data structure that implements thread safe storage of
 // circuits. Each circuit key (payment hash) may have several of circuits
 // corresponding to it due to the possibility of repeated payment hashes.
-//
-// TODO(andrew.shvv) make it persistent
-type circuitMap struct {
-	sync.RWMutex
-	circuits map[circuitKey]*paymentCircuit
+type circuitStorage struct {
+	db *channeldb.DB
 }
 
-// newCircuitMap creates a new instance of the circuitMap.
-func newCircuitMap() *circuitMap {
-	return &circuitMap{
-		circuits: make(map[circuitKey]*paymentCircuit),
-	}
+// newCircuitMap creates a new instance of the circuitStorage.
+func newCircuitMap(db *channeldb.DB) *circuitStorage {
+	return &circuitStorage{db}
 }
 
-// add adds a new active payment circuit to the circuitMap.
-func (m *circuitMap) add(circuit *paymentCircuit) error {
-	m.Lock()
-	defer m.Unlock()
-
-	// Examine the circuit map to see if this circuit is already in use or
-	// not. If so, then we'll simply increment the reference count.
-	// Otherwise, we'll create a new circuit from scratch.
-	//
-	// TODO(roasbeef): include dest+src+amt in key
-	if c, ok := m.circuits[circuit.PaymentHash]; ok {
-		c.RefCount++
-		return nil
-	}
+// add adds a new active payment circuit to the circuitStorage.
+func (m *circuitStorage) add(incomingCircuit *paymentCircuit) error {
+	return m.db.Batch(func(tx *bolt.Tx) error {
+		key := incomingCircuit.Key()
+		circuitBucket, err := tx.CreateBucketIfNotExists(circuitBucketKey)
+		if err != nil {
+			return err
+		}
+
+		counterBucket, err := tx.CreateBucketIfNotExists(counterCircuitBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// Examine the circuit map to see if this circuit is already in use or
+		// not. If so, then we'll simply increment the reference count.
+		// Otherwise, we'll add new circuit in storage.
+		refCounter := uint64(1)
+		if data := counterBucket.Get(key[:]); data != nil {
+			refCounter = binary.BigEndian.Uint64(data)
+			refCounter++
+		}
+
+		// Update reference counter of the payment circuit.
+		var refCounterBytes [8]byte
+		binary.BigEndian.PutUint64(refCounterBytes[:], refCounter)
+		if err := counterBucket.Put(key[:], refCounterBytes[:]); err != nil {
```

Comment:
Just checking: this block of code is threadsafe right? From what I read `db.Batch` should be atomic, but not 100% convinced :)